src/groups/mqb/mqbauthn/mqbauthn_authenticationcontroller.cpp

```diff
@@ -0,0 +1,483 @@
+    enum RcEnum {
+        // Enum for the various RC error categories
+        rc_SUCCESS             = 0,
+        rc_ALREADY_STARTED     = -1,
+        rc_INVALID_CONFIG      = -2,
+        rc_INIT_AUTHENTICATORS = -3
+    };
+
+    if (d_isStarted) {
+        errorDescription << "start() can only be called once on this object";
+        return rc_ALREADY_STARTED;
+    }
+
+    BALL_LOG_INFO << "Starting AuthenticationController";
+
+    int rc = setAnonymousCredential(errorDescription);
+    if (rc != 0) {
+        return rc_INVALID_CONFIG;  // RETURN
+    }
+
+    rc = initializeAuthenticators(errorDescription);
+    if (rc != 0) {
+        return rc * 10 + rc_INIT_AUTHENTICATORS;
+    }
+
+    // Initialize Authenticators from plugins
+
+    d_isStarted = true;
+    return rc_SUCCESS;
+}
+
+void AuthenticationController::stop()
+{
+    if (!d_isStarted) {
+        return;  // RETURN
+    }
+
+    d_isStarted = false;
+
+    // Stop all authenticators
+    for (AuthenticatorMap::iterator it = d_authenticators.begin();
+         it != d_authenticators.end();
+         ++it) {
+        it->second->stop();
+    }
+
+    d_authenticators.clear();
+}
+
+int AuthenticationController::authenticate(
+    bsl::ostream&                                   errorDescription,
+    bsl::shared_ptr<mqbplug::AuthenticationResult>* result,
+    bsl::string_view                                mechanism,
+    const mqbplug::AuthenticationData&              input)
+{
+    // executed by an *AUTHENTICATION* thread
+
+    enum RcEnum {
+        // Enum for the various RC error categories
+        rc_SUCCESS                 = 0,
+        rc_AUTHENTICATION_FAILED   = -1,
+        rc_MECHANISM_NOT_SUPPORTED = -2
+    };
+
+    int                rc = rc_SUCCESS;
+    bmqu::MemOutStream errorStream(d_allocator_p);
+
+    const bsl::string normMech = normalizeMechanism(mechanism, d_allocator_p);
+    AuthenticatorMap::const_iterator cit = d_authenticators.find(normMech);
+    if (cit != d_authenticators.cend()) {
+        const AuthenticatorMp& authenticator = cit->second;
+        BALL_LOG_DEBUG << "AuthenticationController: "
+                       << "authenticating with mechanism '" << normMech << "'"
+                       << " (authenticator: '" << authenticator->name()
+                       << "')";
+        rc = authenticator->authenticate(errorStream, result, input);
+        if (rc != rc_SUCCESS) {
+            errorDescription << "AuthenticationController: failed to "
+                                "authenticate with mechanism '"
+                             << normMech << "'. (rc = " << rc
+                             << "). Detailed error: " << errorStream.str();
+            return (rc * 10 + rc_AUTHENTICATION_FAILED);
+        }
+    }
+    else {
+        errorDescription
+            << "AuthenticationController: authentication mechanism '"
+            << normMech << "' not supported.";
+        return (rc * 10 + rc_MECHANISM_NOT_SUPPORTED);
```

Comment:
`rc == e_SUCCESS` here, might simplify:

```suggestion
        return rc_MECHANISM_NOT_SUPPORTED;
```

We already have functions where we mix `const * rc + enum` with just `rc` or `enum`, this for example: https://github.com/bloomberg/blazingmq/blob/157313364b9eb131df3022f1462834c0f609fcf3/src/groups/mqb/mqbc/mqbc_clusterstateledgerutil.cpp#L512